### ISSUE TITLE
Add DEFLATE decompression to 7z formats

### ISSUE THREAD

**Issue by dosoos** [magnumripper/JohnTheRipper#3476]
## 7z file to john hash, start john cracking will tips:
`Using default input encoding: UTF-8
No password hashes loaded (see FAQ)`

## my 7z hash(is very large, maybe 14MB, hash file still very large 3.8M), convert tips:
`WARNING: to correctly verify the CRC checksum of the data contained within the file '../xxoo.7z',
the data must be decompressed using DEFLATE after the decryption step.

john currently does not support this particular decompression algorithm.

INFO: However there is also some good news in this particular case.
Since AES-CBC is used by the 7z algorithm and the data length of this file allows a padding attack,
the password recovery tool might be able to use that to verify the correctness of password candidates.
By using this attack there might of course be a higher probability of false positives.`

**Comment by kholia** [comment]
### IMPORTANT ###

This is not a support forum, it's a bug tracker. If you don't understand the difference, please do NOT open an issue. For questions and support, review recent postings on the john-users mailing list at https://www.openwall.com/lists/john-users/ and then subscribe to the list at https://www.openwall.com/john/#lists and post your message to the list by e-mailing `john-users at lists.openwall.com`.

We are not interested in bugs in your distro-supplied and/or several years old Jumbo. Please do not open an issue unless you tried the latest version from HERE first.

**Comment by kholia** [comment]
https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/.github/issue_template.md

**Comment by magnumripper** [comment]
Please post the first part of the hash, eg. the output of `cut -c1-100 hashfile`

**Comment by dosoos** [comment]
`jiexiwenjian.7z:$7z$7$19$0$$8$44dc249a38555abb0000000000000000$1767519895$1967632$1967621$43a98f7909`
this is first part of this hash file

**Comment by dosoos** [comment]
@magnumripper ok! i post this hash file first part

**Comment by magnumripper** [comment]
We got an update from upstream. Please update and re-run 7z2john.pl and see if you get a different and working output. Then please report back!

**Comment by magnumripper** [comment]
The problem seems to be the file has a "compression type" of 7 (deflate) and we don't yet support that. I'm not sure why 7z2john.pl doesn't output a "padding attack" hash even though it mentions that, but I believe you can convert it like this:
```
$ sed -ri 's/\$7\$/$128$/' hashfile
```
This should change the first part of the hash from `$7z$7$19$` into `$7z$128$19$` and that will trigger a padding attack. Then John should load it. Please report results.

If you can email me the complete 7z file, I can probably add proper support for it. I think we have all code needed, as well as some ingenious early rejection stuff to speed it up significantly.

**Comment by kholia** [comment]
@magnumripper The following commands can help in reproducing this problem locally.

```
$ dd if=/dev/zero of=sample.txt bs=1M count=32
32+0 records in
32+0 records out
33554432 bytes (34 MB, 32 MiB) copied, 0.0778151 s, 431 MB/s
```

```
$ 7za a -t7z sample.7z sample.txt -m0=Deflate -popenwall
```

```
$ ../run/7z2john.pl sample.7z > hash
WARNING: to correctly verify the CRC checksum of the data contained within the file 'sample.7z',
the data must be decompressed using DEFLATE after the decryption step.

john currently does not support this particular decompression algorithm.

INFO: However there is also some good news in this particular case.
Since AES-CBC is used by the 7z algorithm and the data length of this file allows a padding attack,
the password recovery tool might be able to use that to verify the correctness of password candidates.
By using this attack there might of course be a higher probability of false positives.
```

**Comment by claudioandre-br** [comment]
Bug closed because of:

* **Fixed** - It was fixed.
* **NeedInfo** - represents a bug that cannot be reproduced and fixed until more information is provided.

**Comment by magnumripper** [comment]
Re-opening for adding DEFLATE at some point in the future. I believe we have all code needed, including terrific early-reject logic.

**Comment by magnumripper** [comment]
This was fixed in 106ac04f6
